network/p2p/error.go

```diff
@@ -0,0 +1,33 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import "github.com/ava-labs/avalanchego/snow/engine/common"
+
+var (
+	// ErrUnexpected should be used to indicate that a request failed due to a
+	// generic error
+	ErrUnexpected = &common.AppError{
+		Code:    -1,
+		Message: "unexpected error",
+	}
+	// ErrUnregisteredHandler should be used to indicate that a request failed
+	// due to it not matching a registered handler
+	ErrUnregisteredHandler = &common.AppError{
+		Code:    -2,
+		Message: "unregistered handler",
+	}
+	// ErrNotValidator should be used to indicate that a request failed due to
+	// the requesting peer not being a validator
+	ErrNotValidator = &common.AppError{
+		Code:    -3,
+		Message: "not a validator",
+	}
+	// ErrThrottled should be used to indicate that a request failed due to the
+	// requesting peer exceeding a rate limit
+	ErrThrottled = &common.AppError{
+		Code:    -4,
+		Message: "throttled",
+	}
```

Comment:
Another idea instead of having a separate error code per error type might be to have an error code for a class of errors, analagous to 4xx/5xx errors:

1. Transient client side errors that a client should retry (e.g throttling).
2. Persistent client side errors that a client should not retry (e.g bad request)
3. Transient server side errors that a client should retry (e.g load shed)
4. Persistent server side errors that a client should not retry (e.g internal error)

If we went with this route we could specify the error in the error text, so an error would look like:
```go
AppError{
    Code: -2 // non-retryable client error
    Message: "this request handler is not supported"
}
```


Pros: Don't add an error code for each new error, less error codes we have to maintain.
Cons: Potentially less flexibility to the client. We also are shoe-horned into having each error fit one of these buckets which will be awkward if we run into an unforeseen error class. We can alternatively just have a new error code per error type and expect the client to determine how to handle it best.